[JSC] Stop using Weak for the dependencies in DeferredWorkTimer::TicketData::TicketData

hyjorc1 opened the pull request:
#### 6d458f2e184a8684aee8ead2acc0d255a87045a5
<pre>
[JSC] Stop using Weak for the dependencies in DeferredWorkTimer::TicketData::TicketData
<a href="https://bugs.webkit.org/show_bug.cgi?id=282707">https://bugs.webkit.org/show_bug.cgi?id=282707</a>
<a href="https://rdar.apple.com/139211106">rdar://139211106</a>

Reviewed by Yusuke Suzuki and Keith Miller.

JSC::Weak allocation should not be allowed at GC end phase.
Since DeferredWorkTimer::addPendingWork can be triggered at GC
end phase, we should not use JSC::Weak in the TicketDat. Instead
we can directly manage the pending tickets at GC end phase since
JSC::Weak is just ensuring a null reference once the JSCell is dead.

* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::runEndPhase):
(JSC::Heap::cancelDeferredWorkIfNeeded):
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/runtime/DeferredWorkTimer.cpp:
(JSC::DeferredWorkTimer::TicketData::TicketData):
(JSC::DeferredWorkTimer::TicketData::create):
(JSC::DeferredWorkTimer::TicketData::cancel):
(JSC::DeferredWorkTimer::addPendingWork):
(JSC::DeferredWorkTimer::cancelPendingWork):
* Source/JavaScriptCore/runtime/DeferredWorkTimer.h:
(JSC::DeferredWorkTimer::TicketData::target):
(JSC::DeferredWorkTimer::TicketData::scriptExecutionOwner):
(JSC::DeferredWorkTimer::TicketData::dependencies): Deleted.
(JSC::DeferredWorkTimer::TicketData::globalObject): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::visitChildrenImpl):
* Source/JavaScriptCore/runtime/WaiterListManager.cpp:
(JSC::WaiterListManager::notifyWaiterImpl):
(JSC::WaiterListManager::unregister):
(JSC::Waiter::dump const):
* Source/JavaScriptCore/wasm/WasmStreamingCompiler.cpp:
(JSC::Wasm::StreamingCompiler::StreamingCompiler):
(JSC::Wasm::StreamingCompiler::didComplete):
* Source/JavaScriptCore/wasm/js/JSWebAssembly.cpp:
(JSC::JSWebAssembly::webAssemblyModuleValidateAsync):
(JSC::instantiate):
(JSC::compileAndInstantiate):

Canonical link: <a href="https://commits.webkit.org/286378@main">https://commits.webkit.org/286378@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5661cab169fd589a14b09bda4c021e189a59b4e1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75754 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54780 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28606 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80243 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27020 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63925 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3043 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59417 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17585 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78821 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49310 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65070 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39773 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46709 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22552 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25347 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/68906 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67826 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22891 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81714 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/75017 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3094 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1963 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67640 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3247 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65037 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66942 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16700 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10896 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9029 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/97284 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3052 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21271 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3077 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/4013 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/3084 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/DeferredWorkTimer.cpp:
```diff
@@ -240,6 +237,34 @@ void DeferredWorkTimer::cancelPendingWorkSafe(JSGlobalObject* globalObject)
+    };
+    auto isValid = [&](auto& ticket) {
+        for (JSCell* dependency : ticket->dependencies()) {
+            if (!isLive(dependency))
```
You can just use `vm.heap.isMarked(dependency)`

Constellation commented on Source/JavaScriptCore/runtime/DeferredWorkTimer.cpp:
```diff
@@ -240,6 +237,34 @@ void DeferredWorkTimer::cancelPendingWorkSafe(JSGlobalObject* globalObject)
+            if (!isLive(dependency))
+                return false;
+        }
+        return isLive(ticket->target()->globalObject()) && isLive(ticket->scriptExecutionOwner());
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/DeferredWorkTimer.cpp:
```diff
@@ -65,8 +63,7 @@ inline VM& DeferredWorkTimer::TicketData::vm()
 {
-    m_scriptExecutionOwner.clear();
-    m_globalObject.clear();
+    m_isCancelled = true;
```
Are there any issues if we clear everything here too?
```
m_dependencies = { };
m_scriptExecutionOwner = nullptr;
```

hyjorc1 replied:
We cannot clear `m_dependencies` since it currently
```
// This is const since we don't want to take the lock when visiting TicketData.
const FixedVector<Weak<JSCell>> m_dependencies;
```
This is because `JSGlobalObject::visitChildrenImpl` needs to visit the `m_dependencies`. In order to avoid the potential race condition and deadlock due to the modification during the iteration. We decided to let it be immutable.

hyjorc1 replied:
Discussed offline. We should clear `m_dependencies` when `DeferredWorkTimer::cancelPendingWork(VM& vm)` since it's at GC end phase.

kmiller68 replied:
What's the upside of doing this at GC end? I guess we save a little memory since we don't have to wait for the JSGlobalObject to get swept but it seems like that would be small (~16-32 bytes per ticket) anyway.

hyjorc1 replied:
We think there are two good things:
1. Prevent dangling pointers that point to freed or invalid memory.
2. Improve debugging and testing since it's easier to detect nullptr access.

kmiller68 replied:
Then there should probably be a comment here that none of this is necessary and it's being done for those reasons. Otherwise future folks could think that it's necessary and will have to prove to themselves it's not.

hyjorc1 replied:
Thanks, updated the comment.

kmiller68 commented on Source/JavaScriptCore/runtime/DeferredWorkTimer.cpp:
```diff
@@ -240,6 +244,31 @@ void DeferredWorkTimer::cancelPendingWorkSafe(JSGlobalObject* globalObject)
+        for (JSCell* dependency : ticket->dependencies()) {
+            if (!vm.heap.isMarked(dependency))
+                return false;
+        }
```
I think this check is unnecessary. If `ticket->target()->globalObject()` is marked then these will all be marked because it visits the dependencies. It's probably reasonable to have as an ASSERT though.

hyjorc1 replied:
> If ticket->target()->globalObject() is marked then these will all be marked because it visits the dependencies.

Yes, that should be true. But we still need to check the target. This is because if the target is dead its global object may not be dead.

kmiller68 replied:
The global object marks everything in `m_dependencies` though. So target can't be dead.

hyjorc1 replied:
> The global object marks everything in m_dependencies though. So target can't be dead.

I see. You're right.

hyjorc1 replied:
Updated.

kmiller68 commented on Source/JavaScriptCore/runtime/DeferredWorkTimer.cpp:
```diff
@@ -240,6 +244,31 @@ void DeferredWorkTimer::cancelPendingWorkSafe(JSGlobalObject* globalObject)
+            needToFire = true;
+        }
+    }
+    if (needToFire && !isScheduled() && !m_currentlyRunningTask)
+        setTimeUntilFire(0_s);
```
Why does this need to schedule work? Why can't we just clear the ticket now?

hyjorc1 replied:
The ticket may have its corresponding task pending in the `m_tasks`. Since `DeferredWorkTimer::doWork` handles all cases I think we should just let it remove all related components. I think this makes code clear and consistent.

kmiller68 replied:
If it's in `m_tasks` then it should already be scheduled, I think.

hyjorc1 replied:
Yes, that's why we check `!isScheduled()` here. Added a comment here since we should let `doWork` handle the removal for consistency.

kmiller68 reviewed (changes requested):
hold off on merging until comments are addressed

kmiller68 reviewed (approved):
r=me

hyjorc1 committed:
[JSC] Stop using Weak for the dependencies in DeferredWorkTimer::TicketData::TicketData
https://bugs.webkit.org/show_bug.cgi?id=282707
rdar://139211106

Reviewed by Yusuke Suzuki and Keith Miller.

JSC::Weak allocation should not be allowed at GC end phase.
Since DeferredWorkTimer::addPendingWork can be triggered at GC
end phase, we should not use JSC::Weak in the TicketDat. Instead
we can directly manage the pending tickets at GC end phase since
JSC::Weak is just ensuring a null reference once the JSCell is dead.

* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::runEndPhase):
(JSC::Heap::cancelDeferredWorkIfNeeded):
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/runtime/DeferredWorkTimer.cpp:
(JSC::DeferredWorkTimer::TicketData::TicketData):
(JSC::DeferredWorkTimer::TicketData::create):
(JSC::DeferredWorkTimer::TicketData::cancel):
(JSC::DeferredWorkTimer::addPendingWork):
(JSC::DeferredWorkTimer::cancelPendingWork):
* Source/JavaScriptCore/runtime/DeferredWorkTimer.h:
(JSC::DeferredWorkTimer::TicketData::target):
(JSC::DeferredWorkTimer::TicketData::scriptExecutionOwner):
(JSC::DeferredWorkTimer::TicketData::dependencies): Deleted.
(JSC::DeferredWorkTimer::TicketData::globalObject): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::visitChildrenImpl):
* Source/JavaScriptCore/runtime/WaiterListManager.cpp:
(JSC::WaiterListManager::notifyWaiterImpl):
(JSC::WaiterListManager::unregister):
(JSC::Waiter::dump const):
* Source/JavaScriptCore/wasm/WasmStreamingCompiler.cpp:
(JSC::Wasm::StreamingCompiler::StreamingCompiler):
(JSC::Wasm::StreamingCompiler::didComplete):
* Source/JavaScriptCore/wasm/js/JSWebAssembly.cpp:
(JSC::JSWebAssembly::webAssemblyModuleValidateAsync):
(JSC::instantiate):
(JSC::compileAndInstantiate):

Canonical link: https://commits.webkit.org/286378@main
